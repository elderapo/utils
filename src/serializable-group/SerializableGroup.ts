import { Constructable } from "../types";
import { SerializeGroupTransformationError } from "./serializable-group-errors";

export interface ISerializableGroupOptions<TYPENAME, BASE_CLASS_SHAPE> {
  typename: string;
  BaseClass?: BASE_CLASS_SHAPE;
}

export class DefaultBaseClass {} // @TEMP: https://github.com/microsoft/TypeScript/issues/33023

export class SerializableGroup<PAYLOADS, TYPENAME extends string, BASE_CLASS_SHAPE = Object> {
  private cache = new Map();

  constructor(
    public options: ISerializableGroupOptions<TYPENAME, Constructable<BASE_CLASS_SHAPE>>
  ) {}

  public createClassConstructor<ID extends keyof PAYLOADS, PAYLOAD = PAYLOADS[ID]>(id: ID) {
    const typename = this.options.typename;
    const BaseClass = (this.options.BaseClass || DefaultBaseClass) as Constructable<Object>;

    const Class = class extends BaseClass {
      public readonly id: ID;
      public readonly payload: PAYLOAD;
      public readonly ["___typename___"] = typename;

      constructor(payload: PAYLOAD) {
        super();

        this.id = id;
        this.payload = payload;
      }
    };

    Object.defineProperty(Class, "name", { value: `${typename}<${id}>` });

    this.cache.set(id, Class);

    return Class as Constructable<BASE_CLASS_SHAPE & InstanceType<typeof Class>, [PAYLOAD]>;
  }

  public checkTypename(instance: any): boolean {
    return instance && instance.___typename___ === this.options.typename;
  }

  public serialize(thing: any): any {
    if (!this.checkTypename(thing)) {
      throw new SerializeGroupTransformationError(this, thing);
    }

    const serialized = JSON.parse(JSON.stringify(thing));

    if (thing instanceof Error) {
      serialized.message = thing.message;
    }

    return serialized;
  }

  public deserialize<ID extends keyof PAYLOADS, PAYLOAD = PAYLOADS[ID]>(thing: {
    id: ID;
    payload: PAYLOAD;
    ___typename___: string;
  }) {
    if (!this.checkTypename(thing)) {
      throw new SerializeGroupTransformationError(this, thing);
    }

    return this.recreate(thing.id, thing.payload);
  }

  public recreate<ID extends keyof PAYLOADS, PAYLOAD = PAYLOADS[ID]>(id: ID, payload: PAYLOAD) {
    if (!this.cache.has(id)) {
      throw new SerializeGroupTransformationError(this, { id, payload });
    }

    const Class = this.cache.get(id);

    return new Class(payload) as {
      id: ID;
      payload: PAYLOAD;
      ___typename___: TYPENAME;
    } & BASE_CLASS_SHAPE;
  }
}
