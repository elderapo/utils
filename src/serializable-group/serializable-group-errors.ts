import { CustomError } from "ts-custom-error";
import { SerializableGroup } from "./SerializableGroup";

export class SerializeGroupTransformationError extends CustomError {
  constructor(public sg: SerializableGroup<any, any>, public thing: any) {
    super(
      `SerializeGroup<${sg.options.typename}> couldn't transform Thing<${JSON.stringify(thing)}>!`
    );
  }
}
