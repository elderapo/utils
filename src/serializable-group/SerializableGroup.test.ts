import { SerializableGroup } from "./SerializableGroup";
import { CustomError } from "ts-custom-error";
import { SerializeGroupTransformationError } from "./serializable-group-errors";

describe("SerializableGroup", () => {
  it("should work correctly with Object as BaseClass (not provided)", () => {
    enum IDS {
      ClassA = "CLASS_A"
    }

    type Payloads = {
      [IDS.ClassA]: { a: string };
    };

    const sg = new SerializableGroup<Payloads, "Sample">({
      typename: "Sample"
    });

    const ClassA = sg.createClassConstructor(IDS.ClassA);

    expect(ClassA).toBeTruthy();

    const classAInstance1 = new ClassA({ a: "aaa" });

    expect(classAInstance1).toBeInstanceOf(Object);
    expect(classAInstance1).toBeInstanceOf(ClassA);
    expect(classAInstance1).toMatchSnapshot();

    const classAInstance1Serialized = sg.serialize(classAInstance1);

    expect(classAInstance1Serialized).toMatchSnapshot();

    const classAInstance1Deserialized = sg.deserialize(classAInstance1Serialized);

    expect(classAInstance1Deserialized).toBeInstanceOf(Object);
    expect(classAInstance1Deserialized).toBeInstanceOf(ClassA);
    expect(classAInstance1Deserialized).toMatchSnapshot();
  });

  it("should work correctly with SomeBaseClass as BaseClass", () => {
    enum IDS {
      ClassA = "CLASS_A"
    }

    type Payloads = {
      [IDS.ClassA]: { a: string };
    };

    class SomeBaseClass {
      public getSomething() {
        return "something";
      }
    }

    const sg = new SerializableGroup<Payloads, "Sample", SomeBaseClass>({
      typename: "Sample",
      BaseClass: SomeBaseClass
    });

    const ClassA = sg.createClassConstructor(IDS.ClassA);

    expect(ClassA).toBeTruthy();

    const classAInstance1 = new ClassA({ a: "aaa" });

    expect(classAInstance1).toBeInstanceOf(SomeBaseClass);
    expect(classAInstance1).toBeInstanceOf(ClassA);
    expect(classAInstance1).toMatchSnapshot();

    const classAInstance1Serialized = sg.serialize(classAInstance1);

    expect(classAInstance1Serialized).toMatchSnapshot();

    const classAInstance1Deserialized = sg.deserialize(classAInstance1Serialized);

    expect(classAInstance1Deserialized).toBeInstanceOf(SomeBaseClass);
    expect(classAInstance1Deserialized).toBeInstanceOf(ClassA);
    expect(classAInstance1Deserialized).toMatchSnapshot();

    expect(typeof classAInstance1Deserialized.getSomething).toBe("function");
    expect(classAInstance1Deserialized.getSomething()).toBe("something");
  });

  it("should work correctly with CustomError as BaseClass", () => {
    enum IDS {
      ClassA = "CLASS_A"
    }

    type Payloads = {
      [IDS.ClassA]: { a: string };
    };

    const sg = new SerializableGroup<Payloads, "MyError", CustomError>({
      typename: "MyError",
      BaseClass: CustomError
    });

    const ClassA = sg.createClassConstructor(IDS.ClassA);

    expect(ClassA).toBeTruthy();

    const classAInstance1 = new ClassA({ a: "aaa" });

    expect(classAInstance1).toBeInstanceOf(Error);
    expect(classAInstance1).toBeInstanceOf(CustomError);
    expect(classAInstance1).toBeInstanceOf(ClassA);
    expect(classAInstance1).toMatchSnapshot();

    const classAInstance1Serialized = sg.serialize(classAInstance1);

    expect(classAInstance1Serialized).toMatchSnapshot();

    const classAInstance1Deserialized = sg.deserialize(classAInstance1Serialized);

    expect(classAInstance1Deserialized).toBeInstanceOf(Error);
    expect(classAInstance1Deserialized).toBeInstanceOf(CustomError);
    expect(classAInstance1Deserialized).toBeInstanceOf(ClassA);
    expect(classAInstance1Deserialized).toMatchSnapshot();

    expect(typeof classAInstance1Deserialized.message).toBe("string");
  });

  it("should correctly throw error if serialize is called with wrong ___typename___", () => {
    enum IDS {
      ClassA = "CLASS_A"
    }

    type Payloads = {
      [IDS.ClassA]: { a: string };
    };

    const sg = new SerializableGroup<Payloads, "Sample">({
      typename: "Sample"
    });

    expect(() => sg.serialize({ ___typename___: "aaa" })).toThrowError(
      SerializeGroupTransformationError
    );
    expect(() => sg.serialize({ ___typename___: "Sample" })).not.toThrowError(
      SerializeGroupTransformationError
    );
  });

  it("should correctly throw error if deserialize is called with wrong ___typename___", () => {
    enum IDS {
      ClassA = "CLASS_A"
    }

    type Payloads = {
      [IDS.ClassA]: { a: string };
    };

    const sg = new SerializableGroup<Payloads, "Sample">({
      typename: "Sample"
    });

    const ClassA = sg.createClassConstructor(IDS.ClassA);

    expect(() =>
      sg.deserialize({ id: IDS.ClassA, payload: { a: "aaa" }, ___typename___: "__wrong__" })
    ).toThrowError(SerializeGroupTransformationError);
    expect(() =>
      sg.deserialize({ id: IDS.ClassA, payload: { a: "aaa" }, ___typename___: "Sample" })
    ).not.toThrowError(SerializeGroupTransformationError);
  });

  it("should correctly throw error if recreate is called with wrong id", () => {
    enum IDS {
      ClassA = "CLASS_A"
    }

    type Payloads = {
      [IDS.ClassA]: { a: string };
    };

    const sg = new SerializableGroup<Payloads, "Sample">({
      typename: "Sample"
    });

    const ClassA = sg.createClassConstructor(IDS.ClassA);

    expect(() => sg.recreate("wrongid" as any, { a: "aaa" })).toThrowError(
      SerializeGroupTransformationError
    );
    expect(() => sg.recreate(IDS.ClassA, { a: "aaa" })).not.toThrowError(
      SerializeGroupTransformationError
    );
  });
});
