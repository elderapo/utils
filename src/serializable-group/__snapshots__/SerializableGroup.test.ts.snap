// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SerializableGroup should work correctly with CustomError as BaseClass 1`] = `[MyError<CLASS_A>]`;

exports[`SerializableGroup should work correctly with CustomError as BaseClass 2`] = `
Object {
  "___typename___": "MyError",
  "id": "CLASS_A",
  "message": "",
  "payload": Object {
    "a": "aaa",
  },
}
`;

exports[`SerializableGroup should work correctly with CustomError as BaseClass 3`] = `[MyError<CLASS_A>]`;

exports[`SerializableGroup should work correctly with Object as BaseClass (not provided) 1`] = `
Sample<CLASS_A> {
  "___typename___": "Sample",
  "id": "CLASS_A",
  "payload": Object {
    "a": "aaa",
  },
}
`;

exports[`SerializableGroup should work correctly with Object as BaseClass (not provided) 2`] = `
Object {
  "___typename___": "Sample",
  "id": "CLASS_A",
  "payload": Object {
    "a": "aaa",
  },
}
`;

exports[`SerializableGroup should work correctly with Object as BaseClass (not provided) 3`] = `
Sample<CLASS_A> {
  "___typename___": "Sample",
  "id": "CLASS_A",
  "payload": Object {
    "a": "aaa",
  },
}
`;

exports[`SerializableGroup should work correctly with SomeBaseClass as BaseClass 1`] = `
Sample<CLASS_A> {
  "___typename___": "Sample",
  "id": "CLASS_A",
  "payload": Object {
    "a": "aaa",
  },
}
`;

exports[`SerializableGroup should work correctly with SomeBaseClass as BaseClass 2`] = `
Object {
  "___typename___": "Sample",
  "id": "CLASS_A",
  "payload": Object {
    "a": "aaa",
  },
}
`;

exports[`SerializableGroup should work correctly with SomeBaseClass as BaseClass 3`] = `
Sample<CLASS_A> {
  "___typename___": "Sample",
  "id": "CLASS_A",
  "payload": Object {
    "a": "aaa",
  },
}
`;
